Support passing argc and argv to target program

ChinYikMing opened the pull request:
Half close #150, I left envp in TODO

I assumed all emulator's option are passed before the target program and its args thus I modify the parse_args function consequently

jserv commented on Makefile:
```diff
@@ -9,7 +9,7 @@ CFLAGS += -Wno-unused-label
 
 # Set the default stack pointer
-CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFFF000
+CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFE000 -D DEFAULT_ARGS_ADDR=0xFFFF000
```
Split into a new line. i.e.,
```
CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFE000
CFLAGS += -D DEFAULT_ARGS_ADDR=0xFFFF000
```

jserv commented on src/main.c:
```diff
@@ -103,11 +103,11 @@ static void print_usage(const char *filename)
 }
 
-static bool parse_args(int argc, char **args)
+static bool parse_args(int *argc, char ***args)
```
It is not elegant to have `char ***arg` here? Can you simplify?

jserv reviewed (changes requested):
Resolve the concerns raised by CodeQL.

jserv commented on src/main.c:
```diff
@@ -103,11 +103,11 @@ static void print_usage(const char *filename)
     /* parse each argument in turn */
-    for (int i = 1; i < argc; ++i) {
-        const char *arg = args[i];
+    for (int i = 1; i < *argc; ++i) {
```
You might consider to utilize [getopt(3)](https://man7.org/linux/man-pages/man3/getopt.3.html).
Example: https://github.com/taviso/123elf/blob/main/main.c#L167

ChinYikMing replied:
I think `getopt ` is a cleaner way, but it does not support long option such as --quiet. If we want to parse long option, we might use `getopt_long`  but it is only compliance to GNU ( might having porting problem ), is it still recommended?

jserv replied:
> I think `getopt ` is a cleaner way, but it does not support long option such as --quiet. If we want to parse long option, we might use `getopt_long` but it is only compliance to GNU ( might having porting problem ), is it still recommended?

It is acceptable to handle `--quiet` with GNU extensions since this option is not part of core functionalities. You still have to denote the usage though.

ChinYikMing replied:
Is it also acceptable for other options? Or we could introduce a short option ( such as -q means --quiet ) to replace long option, if so we can use `getopt` seamlessly

jserv replied:
The latter sounds good.

jserv commented on src/riscv.c:
```diff
@@ -124,17 +127,109 @@ void rv_delete(riscv_t *rv)
+                 (void *) (s->mem->mem_base + (uintptr_t) args_p), sizeof(int));
+    args_p++;
+    sp = (uintptr_t *) ((uint32_t *) sp + 1); /* keep argc and args[0] within
+                                                 one word due to RISCV32 ABI */
```
It is meant to be "RV32 ABI" instead of "RISCV32 ABI."

jserv commented on tests/args.c:
```diff
@@ -0,0 +1,10 @@
+#include <stdio.h>
```
Skip this test program. Let's consolidate `tests/readelf/readelf.c` instead.

ChinYikMing committed:
Support passing argc and argv to target program

- half close #150, I left envp in TODO
- refactor parse_args function with getopt to support
  short option which looks cleaner and replace all long
  option with short option
- update README

jserv commented on src/riscv.c:
```diff
@@ -124,17 +127,109 @@ void rv_delete(riscv_t *rv)
+    /* set emulating program's argc and args */
+    /*
+     * store argc, args to state->mem
+     * so that, we can use offset technique to compatible
```
Make the comments grammatically correct.

ChinYikMing committed:
rephrase comments

jserv commented on src/riscv.c:
```diff
@@ -135,6 +138,98 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
+        memory_write(s->mem, (uintptr_t) args_p, (void *) arg,
+                     (args_len + 1) * sizeof(char));
+        args_space[args_space_idx++] = args_len + 1;
+        args_p = (uintptr_t *) (((char *) args_p) + args_len + 1);
```
Rewrite as following:
```c
    args_p = (uintptr_t *) ((uintptr_t) args_p + args_len + 1);
```

jserv commented on src/riscv.c:
```diff
@@ -135,6 +138,98 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
+        args_p = (uintptr_t *) (((char *) args_p) + args_len + 1);
+        args_len_total += args_len + 1;
+    }
+    args_p = (uintptr_t *) (((char *) args_p) - args_len_total);
```
Ditto. Instead of casting to `char *`, specify `uintptr_t`.

ChinYikMing committed:
avoid casting to char *

qwe661234 commented:
In file `src/risc.c`, the `memory_write` has been invoked twice to wrtie argc or args to memory (according to your comments). Why we need to maintain two copies of argc or args in virtual memory?

ChinYikMing commented:
> In file `src/risc.c`, the `memory_write` has been invoked twice to wrtie argc or args to memory (according to your comments). Why we need to maintain two copies of argc or args in virtual memory?

If we are running emulator on 64-bit machine, the virtual address would be 64-bit unless it is compile with -m32 option. Since we are RV32 emulator, so the address space should be < 4GB, from this point we cannot directly access 64-bit address from emulator. Thus, first call of `memory write` store the argc and argv to < 4GB address, and second call of `memory write` for RV32 ABI.

qwe661234 commented:
> If we are running emulator on 64-bit machine, the virtual address would be 64-bit unless it is compile with -m32 option. Since we are RV32 emulator, so the address space should be < 4GB, from this point we cannot directly access 64-bit address from emulator. Thus, first call of `memory write` store the argc and argv to < 4GB address, and second call of `memory write` for RV32 ABI.

Ok, but I still confuse. I guess these two copies are for different purposes. Could you explain the different purpose between store the argc and argv to < 4GB address and `memory write` for RV32 ABI?(or maybe leave some comments) Thx.

ChinYikMing commented:
> Ok, but I still confuse. I guess these two copies are for different purposes. Could you explain the different purpose between store the argc and argv to < 4GB address and `memory write` for RV32 ABI?(or maybe leave some comments) Thx.

For example, we want access the args[0], the instruction might look like this: `lw a0, 4(sp)`. In mapping, emulator will call `memory_read_w`. From implementation of `memory_read_w`, it is base  memory + "offset",  this "offset" should within the memory size otherwise we get segmentation fault. Thus, value stored in `sp+4` should be the offset to the args[0]. The first copy of argc and argv mainly for maintaining the offset and making sure it is < 4GB.

qwe661234 commented:
> For example, we want access the args[0], the instruction might look like this: `lw a0, 4(sp)`. In mapping, emulator will call `memory_read_w`. From implementation of `memory_read_w`, it is base memory + "offset", this "offset" should within the memory size otherwise we get segmentation fault. Thus, value stored in `sp+4` should be the offset to the args[0]. The first copy of argc and argv mainly for maintaining the offset and making sure it is < 4GB.

Does you mean the instruction `lw a0, 4(sp)` get the offset of memory address of args[0], and the data of args[0] store in this memory address(mem_base + offset)?

ChinYikMing commented:
> Does you mean the instruction `lw a0, 4(sp)` get the offset of memory address of args[0], and the data of args[0] store in this memory `address(mem_base + offset)`?

yes

ChinYikMing commented:
> Resolve the concerns raised by CodeQL.

I think I have solved it.

jserv commented:
I defer to @qwe661234 for confirmation.

qwe661234 commented on Makefile:
```diff
@@ -9,7 +9,9 @@ CFLAGS += -Wno-unused-label
 
 # Set the default stack pointer
-CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFFF000
+CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFE000
```
Why you remove the highest 4 bit?

ChinYikMing replied:
> Why you remove the highest 4 bit?

To left some space(4096) for list of arguments. According to the [ARG_MAX](https://man7.org/linux/man-pages/man3/sysconf.3.html), this size of the list of arguments must not be less than _POSIX_ARG_MAX (4096). Since we are emulating target program, so we can decide our own size instead and I think 4096 is decent for now.

qwe661234 replied:
Your reply seems like you change `F000` to `E000`, but my question is why you change `FFFFF000` to `FFFF000`. Maybe I misunderstand your reply, could you explain it more detail.

ChinYikMing replied:
> Your reply seems like you change `F000` to `E000`, but my question is why you change `FFFFF000` to `FFFF000`. Maybe I misunderstand your reply, could you explain it more detail.

My bad, I will change it back to 'FFFFF000'

ChinYikMing committed:
reset to previous DEFAULT_STACK_ADDR

jserv commented:
Thank @ChinYikMing for contributing!
